src/Components/Components/src/ParameterCollectionExtensions.cs

```diff
@@ -41,17 +42,68 @@ public static class ParameterCollectionExtensions
                 _cachedWritersByType[targetType] = writers;
             }
 
+            // The logic is a little convoluted now that we have "Extra" parameters. We want to avoid allocations where
+            // possible.
+            //
+            // Error cases that are possible here:
+            // 1. Using an unknown parameter when there is no "Extra" parameter defined
+            // 2. Using an unknown parameter when there is an "Extra" parameter defined *AND* explicitly
+            //    providing a value for the extra parameter.
+            //
+            // The second case has to be an error because we want to allow users to set the "Extra" parameter
+            // explicitly, but we don't ever want to mutate a value the user gives us. We also don't want to
+            // implicitly copy a value the user gives us. Either one of those implementation choices would do
+            // something unexpected.
+
+            var isExtraParameterSetExplicitly = false;
+            Dictionary<string, object> extras = null;
             foreach (var parameter in parameterCollection)
             {
                 var parameterName = parameter.Name;
-                if (!writers.WritersByName.TryGetValue(parameterName, out var writerWithIndex))
+                if (string.Equals(parameterName, writers.CaptureExtraAttributesPropertyName, StringComparison.OrdinalIgnoreCase))
+                {
+                    isExtraParameterSetExplicitly = true;
+                }
+
+                bool isExtraParameter;
+                if (!writers.WritersByName.TryGetValue(parameterName, out var writer) &&
+                    writers.CaptureExtraAttributesWriter == null)
                 {
+                    // Case 1: There is nowhere to put this value.
                     ThrowForUnknownIncomingParameterName(targetType, parameterName);
+                    throw null; // Unreachable
+                }
+
+                isExtraParameter = writer == null;
+
+                if (isExtraParameter)
+                {
+                    extras ??= new Dictionary<string, object>(StringComparer.OrdinalIgnoreCase);
```

Comment:
`Ordinal`? The behavior's odd even if you do `OrdinalIgnoreCase` since it just knocks out the previous value.

Reply:
Attribute matching in HTML is case-insensitive.